### PR TITLE
Add server-messages module for scheduled broadcasts

### PR DESCRIPTION
**Pull request by niekcandaele**
## Summary

Adds a new Takaro module that broadcasts scheduled announcements to connected players via a cronjob. Operators write a list of messages and pick between sequential rotation (in-order with wrap-around) and a weighted shuffle-bag random rotation. Messages support `{playerCount}` and `{serverName}` placeholders. The cronjob skips ticks without advancing rotation state when no messages are configured or no players are online, and resets state when the effective config changes. Schedule is operator-overridable via Takaro's native `systemConfig.cronJobs.broadcast.temporalValue`, matching how the existing lottery module exposes its schedule.

## Architecture

```
                                     ┌────────────────────────────────────┐
   Takaro cron (default */15 * * * *) │                                    │
            │                         │  Broadcast cronjob                 │
            ▼                         │                                    │
   ┌────────────────┐                 │  1. load config                    │
   │ broadcast      │ ─── reads ──▶   │  2. exit if no messages or         │
   │ cronjob entry  │                 │     no online players              │
   │                │                 │  3. detect config change via       │
   │                │                 │     hash → reset rotation state    │
   │                │                 │  4. pick next message:             │
   │                │                 │       - sequential: sm_seq_index   │
   │                │                 │       - random: sm_bag[cursor]     │
   │                │                 │  5. render placeholders            │
   │                │                 │  6. send to game server            │
   │                │                 │  7. persist next index/cursor      │
   └────────────────┘                 │     ONLY after successful send     │
            │                         └────────────────────────────────────┘
            │
            ├── helpers.js  (variable I/O, config hash, weighted bag, shuffle)
            │
            ▼
   ┌─────────────────────────────────┐
   │ Takaro variableController API   │
   │   sm_config_hash                │
   │   sm_seq_index                  │
   │   sm_bag (text values, cursor)  │
   └─────────────────────────────────┘
```

## What Changed

- **`modules/server-messages/module.json`** — Module manifest. Defines the operator config schema (messages list with `text` + optional `weight`, `order` enum) and the broadcast cronjob with default `*/15 * * * *`. The `text` property has `pattern: "\\S"` so whitespace-only and empty messages are rejected at install time.
- **`modules/server-messages/src/cronjobs/broadcast/index.js`** — The cronjob entry point. Implements: empty-list quiet exit, zero-player skip, config-hash invalidation with first-run vs change distinction, sequential rotation with wrap, weighted shuffle-bag random rotation, `{playerCount}` and `{serverName}` placeholder rendering, allowlist-based unknown-placeholder warn log, duplicate-text warn on state init, sequential and random OOB recovery (reset/rebuild and proceed), and an explicit ordering: rotation state advances **only after** `gameServerControllerSendMessage` succeeds — matching the plan's "skip ticks without advancing state when nothing is actually sent" invariant.
- **`modules/server-messages/src/functions/server-messages-helpers.js`** — Persistence helpers (`getSeqIndex`/`setSeqIndex`, `getBagState`/`setBagState`, `getConfigHash`/`setConfigHash`, `clearRotationState`), bag construction (`buildBag` expands by weight, returns text values), Fisher-Yates shuffle, and a canonical config hash (order-aware: random mode is order-insensitive over messages but does include the `order` field itself, so mode switches correctly invalidate state). `writeVariable` retries on 409 (concurrent first-write race); `deleteVariable` is 404-tolerant for idempotent clears.
- **`modules/server-messages/test/broadcast.test.ts`** — 18 real-API integration tests against the live Takaro dev server (no mocks, per repo testing philosophy). Coverage spans: empty/zero-player skip, sequential rotation + wrap, random bag mechanics (per-cycle counts, exhaustion rebuild, shuffle effectiveness), placeholder rendering, unknown-placeholder warn, both reinstall and in-place config changes, weight changes, sequential-with-weight warn, cosmetic reorder in random mode (no reset, correct text continuation), sequential and random OOB recovery (with post-recovery follow-through assertions), duplicate-text warn in both modes, and mode switches.

## Reviewer Guide

- **Start here**: `modules/server-messages/src/cronjobs/broadcast/index.js` — the cronjob top-to-bottom is the readable narrative. `module.json` first if you want to see the operator surface before the implementation.
- **Pay attention to**:
  - The state-write ordering at lines ~189-194 and ~261-266 (the `setSeqIndex` / `setBagState` calls). The comments there call out a deliberate trade-off: a transient state-write failure after a successful send produces a duplicate broadcast on the next tick. We chose this over the inverse (advance on send failure → silently drop a message) because it matches the plan and matches the afk-kick precedent.
  - The bag stores **message text**, not array indices. This was a turn-4 correctness fix — without it, cosmetic reorder of messages in random mode silently sent the wrong message for the rest of the cycle.
  - `checkUnknownPlaceholders` uses an allowlist (`{playerCount}`, `{serverName}`) so a `{serverName}` that fails to resolve does NOT produce a misleading "unrecognised placeholder" warn. If you add a new placeholder, update the allowlist alongside the renderer.
  - The config-change reset writes the new hash BEFORE clearing rotation state to shrink (not eliminate) the concurrent-execution race window.
- **Design decisions**:
  - Bag stores text instead of indices — chosen over reverting the random-mode canonical hash because it preserves the "cosmetic reorder = no reset" UX.
  - `pattern: "\\S"` on `text` rejects blank messages at install time. We intentionally do not constrain `minLength` further so single-character messages like `"!"` are still allowed.
  - Concurrent-trigger duplicates are accepted as best-effort. Eliminating them would require optimistic concurrency primitives the Takaro variables API does not expose. The OOB guards self-heal within one tick.
  - Module name is `test-server-messages` per the plan and per repo convention (`cleanupTestModules` only sweeps `test-`-prefixed names).

## Implementation Journey

7 turns of an adversarial verify-and-fix loop (severity threshold 3) before approval. See the friction log for the rough patches; the short version is:
- Turn 2 the player tried to advance state on send failure with a wrong justification → reverted in turn 3 to plan-compliant ordering.
- Turn 4 Codex (independent second-model review) caught a stored-index staleness bug nobody else flagged → bag switched to storing text.
- Turn 4 the module name was renamed back from `server-messages` to `test-server-messages` after discovering turn 2's UX-driven rename broke the `cleanupTestModules` safety net.
- Turn 7 final pass deleted the last unreachable defensive block and the loop approved with 0 issues at threshold.

## Friction Log

- **Concurrent-trigger race** (`modules/server-messages/src/cronjobs/broadcast/index.js`): The 409-retry on creates and 404-tolerant deletes narrow the race window but do not close the read-modify-write race on `sm_seq_index` and `sm_bag`. Documented as accepted "best-effort on concurrent fires." OOB guards recover within one tick. Verify the cronjob scheduler does not double-fire in your deployment; if it does, expect occasional duplicate broadcasts.
- **State-write-after-send trade-off** (`modules/server-messages/src/cronjobs/broadcast/index.js:189-194, 261-266`): Send-success-then-state-write-failure causes a duplicate on the next tick. Inline comments at both call sites explain the trade-off. v1 deliberately omits a delivery-receipt pattern.
- **Test wall-clock and external API flakiness** (`modules/server-messages/test/broadcast.test.ts`): The 18 tests take ~500 seconds against live Takaro and full `npm test` runs sometimes hang on external API issues (WS ping timeouts, gameserver hard-deletes on WS churn). Recommended invocation: `node --import=ts-node-maintained/register/esm --test --test-force-exit --test-concurrency 1 modules/server-messages/test/broadcast.test.ts`. The test file has `try/finally` on every test so a single failing assertion does not poison subsequent tests.

## Below-Threshold Issues

- (sev 2) Pre-existing weight=0 TypeError edge case in random mode if every message has `weight: 0` (`buildBag` returns empty → next access throws). Not introduced or masked by this PR; would warrant adding `minimum: 1` on the schema in a follow-up.
- (sev 2) Duplicate-text warn message is on the long side but explains both random-mode and sequential-mode outcomes in one line.

## Testing Plan

> Requires `docker compose up -d paper bot redis` for the integration tests, plus the usual `.env` with Takaro credentials.

### Happy Path
- [ ] Run the isolated test suite: `node --import=ts-node-maintained/register/esm --test --test-force-exit --test-concurrency 1 modules/server-messages/test/broadcast.test.ts` — expect **18/18 pass in ~500s**
- [ ] Push the module: `bash scripts/module-push.sh modules/server-messages` — expect successful push
- [ ] Install with three sequential messages including `{playerCount}` and `{serverName}` placeholders, attach a bot via the bot service, trigger the cronjob, and verify the bot sees the rendered broadcast in chat (e.g. `Online: 1 on nca-ai-paper`)
- [ ] Trigger again and verify sequential rotation advances by index, wrapping at the end of the list

### Edge Cases
- [ ] Install with an unknown placeholder like `{foo}` — verify the broadcast contains the literal `{foo}` and the operator log contains an "unrecognised placeholders" warn naming `{foo}` (and **not** `{serverName}` or `{playerCount}`)
- [ ] Disconnect the bot, trigger — verify the operator log says `"no players online, skipping tick without advancing state"` and the next reconnected trigger picks up at the same index (state was not consumed)
- [ ] Try to install with a whitespace-only or empty message text — expect a 400 ValidationError citing `pattern: "\\S"`
- [ ] Install in random mode with three distinct messages, consume one or two via triggers, then reinstall with the same messages in a different array order — verify NO `"config changed"` log and the remaining bag entries are sent (correct text per remaining slots)
- [ ] Install in random mode with `[{text:'A'}, {text:'B'}, {text:'A'}]` and trigger — verify the duplicate-text warn fires once on init and explains the combined-weight outcome for both modes
- [ ] Switch a running install from `order: sequential` to `order: random` (same messages) — verify the config-change reset log fires (`"config changed (messages or order)..."`)
- [ ] Trigger 3-4 times in quick succession — verify no crashes; some duplicate broadcasts are acceptable per the documented best-effort concurrent behavior

### Regression Checks
- [ ] Existing afk-kick / lottery / community-fund / kill-tracker modules continue to install and trigger correctly (this PR adds a new module and does not modify shared helpers)
- [ ] `scripts/module-push.sh` and `scripts/module-pull.sh` continue to work for other modules
- [ ] `cleanupTestModules` still cleans up the `test-server-messages` module on test teardown (the `test-` prefix is preserved deliberately)

### CODE REVIEW

File: modules/server-messages/src/cronjobs/broadcast/index.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+      opts: {},
+    });
+
+    // If sendMessage succeeds but this state-write fails, the exception propagates
+    // and the cronjob is marked failed. The next tick will resend this message
+    // (double-broadcast). This is an inherent trade-off of the "no-advance-on-send-failure"
+    // invariant per the plan — a transient storage failure after a successful send is
+    // rare but causes a visible duplicate. Delivery-receipt / pending-marker patterns
+    // could eliminate this but are deliberately out of scope for v1.
+    await setSeqIndex(gameServerId, moduleId, nextIndex);
```
Comment (niekcandaele) [inline modules/server-messages/src/cronjobs/broadcast/index.js:195]: Deliberate ordering: state advances ONLY after `gameServerControllerSendMessage` succeeds. If the send throws, the cronjob is marked failed and the next tick retries the same message. The inverse failure mode (send succeeds but this state-write throws) produces a duplicate broadcast on retry — accepted as a v1 trade-off; closing it would need a delivery-receipt pattern. Same setup at line ~260 for random mode.

File: modules/server-messages/src/functions/server-messages-helpers.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+  return JSON.stringify({ order, messages: normalized });
+}
+
+/**
+ * Build a weighted bag of message texts.
+ * Each message's text appears (weight) times in the bag.
+ * Storing text (not array indices) means a cosmetic reorder of config messages
+ * does not corrupt the bag — the same text still refers to the same logical message.
+ */
+export function buildBag(messages) {
```
Comment (niekcandaele) [inline modules/server-messages/src/functions/server-messages-helpers.js:111]: The bag stores message **text**, not array indices. This was a turn-4 correctness fix — without it, cosmetic reorder of messages in random mode silently sent the wrong message for the rest of the cycle. `messages.findIndex(m => m.text === messageText)` is used at the call site to map back to a position when needed.

File: modules/server-messages/src/functions/server-messages-helpers.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+ * Write a module-scoped variable (upsert).
+ * Handles 409 conflict (race condition on concurrent first-write) by reading
+ * the winning writer's value and returning it, so callers can continue safely.
+ *
+ * The return value on a 409 is informational — it contains the winning writer's
+ * stored value. Callers that only perform a write (and don't need to read back
+ * the stored value) can safely ignore it; the next read will reflect Takaro's
+ * state. Callers that need the authoritative value should use it.
+ */
+async function writeVariable(gameServerId, moduleId, key, value) {
```
Comment (niekcandaele) [inline modules/server-messages/src/functions/server-messages-helpers.js:31]: Concurrency mitigation: 409 (write race on first creation) is caught and re-reads the winner's value. Other status codes propagate. Combined with `deleteVariable`'s 404 tolerance below, this narrows but does not close the read-modify-write race on rotation state. Documented as best-effort — Takaro's cronjob scheduler is expected not to double-fire and the OOB guards self-heal within one tick.

File: modules/server-messages/src/cronjobs/broadcast/index.js [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    console.log(`server-messages: random mode, built new bag of size ${newBag.length}`);
+  }
+
+  let messageText = bagState.bag[bagState.cursor];
+
+  // OOB guard: if the stored text is no longer in the current message list (shouldn't
+  // happen post-reset but possible if state is corrupt or messages were removed without
+  // a hash change), rebuild the bag and proceed from position 0.
+  const currentTexts = new Set(messages.map((m) => m.text));
+  if (messageText === undefined || !currentTexts.has(messageText)) {
```
Comment (niekcandaele) [inline modules/server-messages/src/cronjobs/broadcast/index.js:222]: OOB guard for random mode: if the persisted bag references a text no longer in the current message list, rebuild from scratch. This is the recovery path for stale state surviving a non-detected config change (in practice should never trigger because the config hash includes message texts and order).

**Review by niekcandaele — COMMENTED** [review]
Self-review: a few areas where it helps to know the trade-off before reading the code.
